gem/optimise.py

```diff
@@ -413,3 +475,555 @@ def aggressive_unroll(expression):
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return True
+    return False
+
+
+def _list_2_node(children, self):
+    if not self.balanced:
+        return reduce(self.func, children, self.base)
+    if len(children) < 3:
+        return reduce(self.func, children, self.base)
+    else:
+        mid = len(children) // 2
+        return self((self(children[:mid]), self(children[mid:])))
+
+
+def list_2_node(function, children, balanced=True, sort=False):
+    """
+    generate gem node from list of children
+    use recursion so that each term is not too long
+    i.e. (a+b) + (c+d) instead of (((a+b)+c)+d
+    :param children: list of children nodes
+    :return: gem node
+    """
+    # TODO: DAG awareness. Hashing with tuple is probably slow here.
+    if sort:
+        children = sorted(children)
+    if function == Sum:
+        base = Zero()
+    elif function == Product:
+        base = one
+    else:
+        raise AssertionError('Cannot combine unless Sum or Product')
+    mapper = Memoizer(_list_2_node)
+    mapper.func = function
+    mapper.base = base
+    mapper.balanced = balanced
+    return mapper(tuple(children))
+
+
+def sort_keys(keys, arg_ind_flat):
+    """
+    sort keys of :class: `Summand` by lexicographical order, according to the order they
+    appear in :param arg_ind_flat.
+    e.g. [j, k, (j,k)] => [j, (j,k), k]
+    """
+    # Convert keys to string and sort by lexicographical order
+    index_2_letter = dict().fromkeys(arg_ind_flat)
+    start_number = ord('a')
+    for number, arg_ind in enumerate(arg_ind_flat, start=start_number):
+        index_2_letter[arg_ind] = chr(number)
+    strings = dict().fromkeys(keys)
+    for key in keys:
+        if isinstance(key, Index):
+            strings[key] = [index_2_letter[key]]
+        else:
+            strings[key] = [index_2_letter[i] for i in key]
+    return sorted(keys, key=lambda k: strings[k])
+
+
+class Summand(OrderedDict):
+    """
+    An object wrapped around an OrderedDict to represent product of gem nodes (factors)
+    The factors are indexed with keys:
+        1. One of the argument indices (e.g. j):
+            factors which depend on j but not other argument indices
+            All summands will have this item (to avoid checking existence),
+            with the values being an empty list possibly.
+        2. Tuple of more than one argument indices (e.g. (j, k)):
+            factors which depend on j and k, but not other argument indices
+            These are rarer and are added as factors are encourtered
+        3. string 'const':
+            factors with no free indices
+        4. string 'other':
+            factors depend on indices (typically quadrature indices for reduction)
+            other than argument indices
+    """
+    def __init__(self, *args, **kwargs):
+        OrderedDict.__init__(self, *args, **kwargs)
+        if 'arg_ind_flat' in kwargs:
+            self.pop('arg_ind_flat')
```

Comment:
This is horrible, but let me see how it is used before I could suggest alternatives.